### PR TITLE
Migrate tslint to eslint

### LINKED ISSUE #689: Since tslint is deprecated, should we switch to eslint?

### PR DESCRIPTION
**Pull request by dimabory**
# All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
- [x] Have you written unit tests?
- [x] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
- [x] This PR is associated with an existing issue?

**Closing issues**

closes #689 

### If this is a new feature submission:

- [x] Has the issue had a maintainer respond to the issue and clarify that the feature is something that aligns with the [goals](https://github.com/lukeautry/tsoa#goal) and [philosophy](https://github.com/lukeautry/tsoa#philosophy) of the project?

**Potential Problems With The Approach**

```
5 ESLint rules behave differently from their TSLint counterparts:
  * @typescript-eslint/no-unused-expressions:
    - The TSLint optional config "allow-new" is the default ESLint behavior and will no longer be ignored.
  * eqeqeq:
    - Option "smart" allows for comparing two literal values, evaluating the value of typeof and null comparisons.
  * no-invalid-this:
    - Functions in methods will no longer be ignored.
  * no-underscore-dangle:
    - Leading and trailing underscores (_) on identifiers will now be ignored.
  * prefer-arrow/prefer-arrow-functions:
    - ESLint does not support allowing standalone function declarations.
    - ESLint does not support allowing named functions defined with the function keyword.
```

**Test plan**

✅ Running `npm run lint` should have the same result as before with `tslint` installed.
✅ No BC
✅ Tests passing

### CODE REVIEW

File: tslint.json [hunk 1]
```diff
@@ -1,25 +0,0 @@
-{
-  "extends": ["tslint:recommended"],
-  "rules": {
-    "array-type": [true, "array-simple"],
```
Comment (dimabory) [inline tslint.json:4]: covered by https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR15-R20

File: tslint.json [hunk 2]
```diff
@@ -1,25 +0,0 @@
...
-    "arrow-parens": false,
```
Comment (dimabory) [inline tslint.json:5]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR92

File: tslint.json [hunk 3]
```diff
@@ -1,25 +0,0 @@
...
-    "object-literal-sort-keys": false,
-    "semicolon": false,
-    "trailing-comma": false,
-    "quotemark": false,
-    "max-line-length": false,
-    "align": false,
-    "no-unused-expression": false,
```
Comment (dimabory) [inline tslint.json:12]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR64

Comment (dimabory) [inline tslint.json:12]: off by default

File: tslint.json [hunk 4]
```diff
@@ -1,25 +0,0 @@
...
-  "extends": ["tslint:recommended"],
-  "rules": {
-    "array-type": [true, "array-simple"],
-    "arrow-parens": false,
-    "object-literal-sort-keys": false,
-    "semicolon": false,
-    "trailing-comma": false,
-    "quotemark": false,
-    "max-line-length": false,
-    "align": false,
```
Comment (dimabory) [inline tslint.json:11]: off by default

File: tslint.json [hunk 5]
```diff
@@ -1,25 +0,0 @@
...
-    "no-unused-expression": false,
-    "no-shadowed-variable": false,
-    "interface-name": [true, "never-prefix"],
-    "member-ordering": [false],
```
Comment (dimabory) [inline tslint.json:15]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR40

File: tslint.json [hunk 6]
```diff
@@ -1,25 +0,0 @@
...
-    "max-classes-per-file": false,
```
Comment (dimabory) [inline tslint.json:16]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR98

File: tslint.json [hunk 7]
```diff
@@ -1,25 +0,0 @@
...
-    "no-reference": false,
-    "no-namespace": [false],
```
Comment (dimabory) [inline tslint.json:18]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR54

File: tslint.json [hunk 8]
```diff
@@ -1,25 +0,0 @@
...
-    "object-literal-key-quotes": false,
-    "ban-types": false
```
Comment (dimabory) [inline tslint.json:20]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR23

File: tslint.json [hunk 9]
```diff
@@ -1,25 +0,0 @@
...
-    "max-line-length": false,
-    "align": false,
-    "no-unused-expression": false,
-    "no-shadowed-variable": false,
-    "interface-name": [true, "never-prefix"],
-    "member-ordering": [false],
-    "max-classes-per-file": false,
-    "no-reference": false,
-    "no-namespace": [false],
-    "object-literal-key-quotes": false,
```
Comment (dimabory) [inline tslint.json:19]: off by default

File: tslint.json [hunk 10]
```diff
@@ -1,25 +0,0 @@
...
-    "trailing-comma": false,
-    "quotemark": false,
-    "max-line-length": false,
-    "align": false,
-    "no-unused-expression": false,
-    "no-shadowed-variable": false,
-    "interface-name": [true, "never-prefix"],
-    "member-ordering": [false],
-    "max-classes-per-file": false,
-    "no-reference": false,
```
Comment (dimabory) [inline tslint.json:17]: no equivalent

File: tslint.json [hunk 11]
```diff
@@ -1,25 +0,0 @@
...
-    "arrow-parens": false,
-    "object-literal-sort-keys": false,
-    "semicolon": false,
-    "trailing-comma": false,
-    "quotemark": false,
-    "max-line-length": false,
-    "align": false,
-    "no-unused-expression": false,
-    "no-shadowed-variable": false,
-    "interface-name": [true, "never-prefix"],
```
Comment (dimabory) [inline tslint.json:14]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR41-R51

File: tslint.json [hunk 12]
```diff
@@ -1,25 +0,0 @@
...
-    "array-type": [true, "array-simple"],
-    "arrow-parens": false,
-    "object-literal-sort-keys": false,
-    "semicolon": false,
-    "trailing-comma": false,
-    "quotemark": false,
-    "max-line-length": false,
-    "align": false,
-    "no-unused-expression": false,
-    "no-shadowed-variable": false,
```
Comment (dimabory) [inline tslint.json:13]: https://github.com/lukeautry/tsoa/pull/780/files#diff-e4403a877d80de653400d88d85e4801aR122-R127

File: tslint-to-eslint-config.log [hunk 13]
```diff
@@ -0,0 +1,13 @@
+5 ESLint rules behave differently from their TSLint counterparts:
```
Comment (WoH) [inline tslint-to-eslint-config.log:1]: Remove

File: packages/cli/src/swagger/specGenerator2.ts [hunk 14]
```diff
@@ -268,7 +268,7 @@ export class SpecGenerator2 extends SpecGenerator {
     }
 
     if (parameter.in === 'query' && parameterType.type === 'array') {
-      (parameter as Swagger.QueryParameter).collectionFormat = 'multi';
```
Comment (WoH) [inline packages/cli/src/swagger/specGenerator2.ts:271]: Why is this cast removed?

File: packages/cli/src/metadataGeneration/typeResolver.ts [hunk 15]
```diff
@@ -883,7 +883,7 @@ export class TypeResolver {
 
   private getModelAdditionalProperties(node: UsableDeclaration) {
     if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
-      const interfaceDeclaration = node as ts.InterfaceDeclaration;
+      const interfaceDeclaration = node;
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/typeResolver.ts:886]: Keep the case

File: packages/cli/src/metadataGeneration/typeResolver.ts [hunk 16]
```diff
@@ -739,7 +739,7 @@ export class TypeResolver {
     const leftmostIdentifier = this.resolveLeftmostIdentifier(type);
     const statements: any[] = this.resolveModelTypeScope(leftmostIdentifier, this.current.nodes);
 
-    const typeName = type.kind === ts.SyntaxKind.Identifier ? (type as ts.Identifier).text : (type as ts.QualifiedName).right.text;
+    const typeName = type.kind === ts.SyntaxKind.Identifier ? type.text : type.right.text;
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/typeResolver.ts:742]: Keep the cast

File: packages/cli/src/metadataGeneration/typeResolver.ts [hunk 17]
```diff
@@ -695,14 +695,14 @@ export class TypeResolver {
...
-      type = (type as ts.QualifiedName).left;
+      type = type.left;
     }
-    return type as ts.Identifier;
+    return type;
   }
 
   private resolveModelTypeScope(leftmost: ts.EntityName, statements: any): any[] {
     while (leftmost.parent && leftmost.parent.kind === ts.SyntaxKind.QualifiedName) {
-      const leftmostName = leftmost.kind === ts.SyntaxKind.Identifier ? (leftmost as ts.Identifier).text : (leftmost as ts.QualifiedName).right.text;
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/typeResolver.ts:705]: Keep the cast

File: packages/cli/src/metadataGeneration/typeResolver.ts [hunk 18]
```diff
@@ -695,14 +695,14 @@ export class TypeResolver {
 
   private resolveLeftmostIdentifier(type: ts.EntityName): ts.Identifier {
     while (type.kind !== ts.SyntaxKind.Identifier) {
-      type = (type as ts.QualifiedName).left;
+      type = type.left;
     }
-    return type as ts.Identifier;
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/typeResolver.ts:700]: Keep the cast

File: packages/cli/src/metadataGeneration/typeResolver.ts [hunk 19]
```diff
@@ -695,14 +695,14 @@ export class TypeResolver {
 
   private resolveLeftmostIdentifier(type: ts.EntityName): ts.Identifier {
     while (type.kind !== ts.SyntaxKind.Identifier) {
-      type = (type as ts.QualifiedName).left;
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/typeResolver.ts:698]: Keep the cast

File: packages/cli/src/metadataGeneration/initializer-value.ts [hunk 20]
```diff
@@ -6,7 +6,7 @@ export const getInitializerValue = (initializer?: ts.Expression, typeChecker?: t
     return;
   }
 
-  switch (initializer.kind as ts.SyntaxKind) {
+  switch (initializer.kind) {
```
Comment (WoH) [inline packages/cli/src/metadataGeneration/initializer-value.ts:9]: Keep the cast

**Review by WoH — COMMENTED** [review]
Alright, I worked through most of the changes and seems great so far, thanks for moving this along.
One question: the eslintrc seems kinda bloated, is there a way to remove rules which are erroring in the preset already?

File: tslint-to-eslint-config.log (hunk 13 above)
Comment (dimabory) [reply tslint-to-eslint-config.log]: good catch

**Comment by dimabory** [comment]
All right @WoH, I cleaned up the config.

**Review by WoH — APPROVED** [review]
Thank you so much, LGTM

**Pull request merged**